src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1068 @@
+      // TBD: should we through an exception ??
+      return;
+    }
+
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
+      // TBD: should we through an exception ??
+      return;
+    }
+    // generate an empty epoch with zero counters
+    int ret = set_epoch(ioctx, d_cluster_id, dpp, 0, 0);
+    if (ret != 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::failed set_epoch()! ret="
+			<< ret << "::" << cpp_strerror(ret) << dendl;
+      return;
+    }
+
+    dedup_epoch_t epoch;
+    reset(store, ioctx, &epoch, 0, 0);
+  }
+
+  //---------------------------------------------------------------------------
+  int cluster::reset(rgw::sal::RadosStore *store,
+		     librados::IoCtx &ioctx,
+		     dedup_epoch_t *p_epoch,
+		     work_shard_t num_work_shards,
+		     md5_shard_t num_md5_shards)
+  {
+    ldpp_dout(dpp, 10) << __func__ << "::REQ num_work_shards=" << num_work_shards
+		       << "::num_md5_shards=" << num_md5_shards << dendl;
+    clear();
+
+    while (true) {
+      int ret = get_epoch(ioctx, dpp, p_epoch, __func__);
+      if (ret != 0) {
+	return ret;
+      }
+      if (p_epoch->num_work_shards && p_epoch->num_md5_shards) {
+	ldpp_dout(dpp, 10) << __func__ << "::ACC num_work_shards=" << p_epoch->num_work_shards
+			   << "::num_md5_shards=" << p_epoch->num_md5_shards << dendl;
+	break;
+      }
+      else if (!num_work_shards && !num_md5_shards) {
+	ldpp_dout(dpp, 10) << __func__ << "::Init flow, no need to wait" << dendl;
+	break;
+      }
+      else {
+	ret = swap_epoch(dpp, ioctx, p_epoch, p_epoch->dedup_type,
+			 num_work_shards, num_md5_shards);
+      }
+    }
+
+    d_epoch_time = p_epoch->time;
+    cleanup_prev_run(ioctx);
+    create_shard_tokens(ioctx, p_epoch->num_work_shards, WORKER_SHARD_PREFIX);
+    create_shard_tokens(ioctx, p_epoch->num_md5_shards, MD5_SHARD_PREFIX);
+
+    return 0;
+  }
+
+  //---------------------------------------------------------------------------
+  int cluster::cleanup_prev_run(librados::IoCtx &ioctx)
+  {
+    constexpr uint32_t max = 100;
+    std::string marker;
+    bool truncated = false;
+    rgw::AccessListFilter filter{};
+    do {
+      std::vector<std::string> oids;
+      int ret = rgw_list_pool(dpp, ioctx, max, filter, marker, &oids, &truncated);
+      if (ret == -ENOENT) {
+	ret = 0;
+	ldpp_dout(dpp, 10) << __func__ << "::rgw_list_pool() ret == -ENOENT"<< dendl;
+	break;
+      }
+      else if (ret < 0) {
+	ldpp_dout(dpp, 1) << "failed rgw_list_pool()! ret=" << ret
+			  << "::" << cpp_strerror(ret) << dendl;
+	return ret;
+      }
+      unsigned deleted_count = 0, skipped_count = 0, failed_count= 0;;
+      for (const std::string& oid : oids) {
+	if (oid == DEDUP_WATCH_OBJ || oid == DEDUP_EPOCH_TOKEN) {
+	  ldpp_dout(dpp, 10) << __func__ << "::skipping " << oid << dendl;
+	  skipped_count++;
+	  continue;
+	}
+	uint64_t size;
+	struct timespec tspec;
+	if (ioctx.stat2(oid, &size, &tspec) != 0) {
```

Comment:
if ret is `-ENOENT` this is probably ok

Reply:
could this happen if 2 separate cleanup commands are called at the same time?

Reply:
okay